spec: enable/disable persistence (#33)
tilt-dev/tilt.specs: enable_disable_persistence.md

# Enable/Disable State Persistence Spec

## Background

Tilt currently has a system for changing the arguments to `tilt up` at
runtime. We usually call it `tilt args`.

https://docs.tilt.dev/tiltfile_config.html

Most teams that use this system are using it for enabling and disabling sets of
services, as demonstrated in this example:

https://docs.tilt.dev/tiltfile_config.html#run-a-defined-set-of-services

This helps save CPU when you only need to run a subset of your services.

A number of teams have built wrappers around `tilt args`. This has helped us to
learn a lot about the ergonomics that people want, and the pains of the current
system. We have mocks and a pretty good idea of how we want the narrowly-focused
single-click enable/disable interaction to work.

For more details, see [Enable/Disable Resource API Spec](./enable_disable_resources.md).

But we have different ideas on how we want the enable/disable interaction to work

> across multiple Tilt runs, and how the state should persist! [^1]

Author: yes

> ## Primary Goal

Author: 1. I think it's totally on the table to either change the semantics of `tilt args`, or to introduce a new CLI that accomplishes similar product goals (e.g., "start tilt with project1", "start tilt with project1+3 extra services", etc)

2. Ya, I also outlined the semantics of systemd below. But I feel like saying "the semantics of docker-compose up/down would make this implementation simpler" is a bit like saying "if we rewrote Tilt in Javascript, we wouldn't have panics anymore." That's true, it would solve that problem! But I think it would also create tons of new problems (as much as i like JS)

4. Click "disable" on "Paulie" -> Roberta is enabled, Joe and Paulie are disabled.
5. Ctrl-c and re-run `tilt up paulie roberta` -> Paulie and Roberta are enabled, Joe is disabled.

Possible Principle:

Tilt always respects arguments to `tilt up` when it starts.

### Specify services, Enable from UI

1. Write a Tiltfile with services Paulie and Roberta and Joe
2. Run `tilt up paulie roberta` -> Paulie and Roberta are enabled, Joe is disabled.
4. Click "enable" on "Joe" -> Paulie and Roberta and Joe are enabled.
5. Ctrl-c and re-run `tilt up paulie roberta` -> Paulie and Roberta are enabled, Joe is disabled. 

Possible Principle:

Tilt always respects arguments to `tilt up` when it starts.

Your UI change has been wiped out by the `tilt up` args.

### Run all, Disable from UI, Change Args

1. Write a Tiltfile with services Paulie and Roberta
2. Run `tilt up` -> Paulie and Roberta are enabled.
4. Click "disable" on "Paulie" -> Roberta is enabled, Paulie is disabled.
5. Run `tilt args paulie roberta` in a separate terminal. -> Tiltfile re-executes, Paulie and Roberta are enabled. 

Possible Principle:

Changing the tilt args wipes out any enabled/disabled services from the UI.

> ### Run one service, Enable from UI, Change Args

Reviewer: I agree with all the other specs but this one feels too much like an exception. I'd prefer to `tilt args` to do the same things every time. (So in this case running `tilt args paulie` the second time would make it so that only `paulie` is enabled.

Author: Ya. This one comes up A LOT in kubernetes-land. 

The classic example is `kubectl apply` - if the input doesn't change, it does nothing on the second apply, but lots of people expect it to restart pods.

we probably could make this work but it'd be messy.

1. Write a Tiltfile with services Paulie and Roberta
2. Run `tilt up paulie` -> Paulie is enabled, Roberta is disabled.
4. Click "enable" on "Roberta" -> Paulie and Roberts is enabled.
5. Run `tilt args paulie` in a separate terminal. -> Tiltfile re-executes, Paulie and Roberta stay enabled.
6. Run `tilt args roberta` in a separate terminal. -> Tiltfile re-executes, Roberta is enabled, Paulie is disabled.

Possible Principle:

Changing the tilt args wipes out any enabled/disabled services from the UI. But
keeping the tilt args the same does nothing.

> 

Reviewer 2: I think we need some examples w/ `config.set_enabled_services` - these are the cases that are tripping me up and making me consider separating service selection from tilt args.
e.g.:
```
services = config.define_string_list('services', args=True)
cfg = config.parse()
services = ['roberta'] + cfg['services']
config.set_enabled_services(services)
```
and then:
1. `tilt up paulie` -> Roberta and Paulie are enabled
2. Disable Roberta from the UI.
3. Run `tilt up juliana`. The user specified 'juliana', but the `config.set_enabled_resources` call got 'juliana' and 'roberta'

A CLI that gets directly at services (or groups) and allows adding/removing rather than replacing seems like it simplifies this a lot. It seems worth looking into how much our users actually want to do arbitrarily complex things with args -> services vs just directly supporting the enabling/disabling of services and groups of services.

Author: sure, added a `set_enabled_resources` case, I'm not sure it makes the implementation any more complex but maybe i'm missing something

system. We have mocks and a pretty good idea of how we want the narrowly-focused
single-click enable/disable interaction to work.

For more details, see [Enable/Disable Resource API Spec](./enable_disable_resource.md).

But we have different ideas on how we want the enable/disable interaction to work
across multiple Tilt runs, and how the state should persist! [^1]

## Primary Goal

The goal of this doc is to give a product specification by example: describe
specific sequences of user actions, and how we want Tilt to behave. (Note that
it is not intended as a description of how Tilt behaves today.)

## Non-Goals

### Groups

We expect that we will eventually want to control groups of services, but
we don't believe that changes our analysis here about how single services behave.

## Terminology

### Enabled

We say a server is "enabled" if it's visible in the Tilt UI and is in status
waiting on trigger, pending on dependency, running, completed, or error.

### Disabled

We say a server is "disabled" if it's visible in the Tilt UI and is in status
"disabled". We do not expect any new logs or status changes in the UI until the

> service is intentionally "enabled"

Author: my current understanding of the decision is that we want tilt to know about all services in your tiltfile (so you can list them, search them, etc). And that hiding them is a purely UI-level action. Though Surbhi / L can add more color on this.

> ### Disabled
> 
> We say a server is "disabled" if it's visible in the Tilt UI and is in status
> "disabled". We do not expect any new logs or status changes in the UI until the
> service is intentionally "enabled"

Reviewer 3: to get at both of these comments and the spec description, i think that the question of whether or not disabled services are visible to the user and to what extent they're visible (through the web app or through the cli) is a design and ux question that can be considered separate from how we define a "disabled" service. 

the reason i say that is because the design could change, but the state of the resource doesn't, ie, a disabled resource is one where the resource status is the equivalent of doing `tilt down <this resource>`. (i may not be using all the right terminology here, so lmk if i'm not making sense!) 

i'd also +1 for resolving the conflicting definitions of a disabled service, even if that's just internal for the moment and doesn't result in any code changes.

here's what i mean by removing the ui definition:
```suggestion
### Disabled

We say a server is "disabled" if it's described by the Tilt API and its status
"disabled". We do not expect any new logs or status changes until the
service is intentionally "enabled"
```

> service is intentionally "enabled"

Author: ya, i like that!

### Session

We say a Tilt "session" is the time from when the user runs "tilt up" to when
the tilt process exits.

### Service Names

For readability, we'll use NY pizza places as service names (Paulie, Juliana, Roberta, Joe).

> ## Prior Art

Author: yes. here's a good doc: http://engblog.yext.com/edward/commands/

loads/reloads, the trigger mode reverts to what's specified in the Tiltfile.

### systemd

[systemd and
systemctl](https://www.commandlinux.com/man-page/man1/systemctl.1.html) is
probably the most widely known service manager. You tend to see its verbs in
other service management tools (like
[Edward](http://engblog.yext.com/edward/commands/#start)).

`systemd` is a persistent daemon that runs admin-level services on linux.

You use `systemctl start` and `systemctl stop` to start/stop services.

And `systemctl enable` and `systemctl disable` to enable/disable services to
auto-boot at system startup.

Note that `enable` does not start a service immediately. In other words, systemd
models "start on boot" and "start now" as independent actions.

## Product Specification

### Run all, Disable from UI

1. Write a Tiltfile with services Paulie and Roberta
1. Run `tilt up` -> Paulie and Roberta are enabled.
1. Click "disable" on "Paulie" -> Roberta is enabled, Paulie is disabled
1. Ctrl-c and re-run `tilt up` -> Roberta is enabled, Paulie is disabled

Possible Principle:

`tilt up` (without arguments) always inherits all the enable/disable overides

> from the previous Tilt session.

Reviewer 3: i could see this perhaps being unexpected or confusing behavior as a principle. (but i could be wrong!) it could be nice functionality to have as an opt-in flag, like `tilt up --use-last-run`.

for me, i think there is something intuitive in the difference between:
- tiltfile: these are my saved settings
- tilt up: this is how i'm configuring this session, but don't persist across sessions
- dashboard/cli: these settings are changed within a session, but don't persist across sessions

Author: L owns this decision, and is responsible for collecting feedback on whether it's the right one. If we get new info on this, you should raise that with them. 

The reason I wrote this doc is because they said, "hey Nick, I've decided that 'tilt up' should persist enable/disable state, can you write up a doc detailing all the implications of that?" Which is here! 

I have my own product opinions on this, but let's put them aside. If I was putting my product critique hat on your proposal, I'd raise an eyebrow at this statement:

> tilt up: this is how i'm configuring this session

If that's your user story, then it seems like you'd want the configuration parameters to be mandatory? i.e., 'tilt up' with no arguments would be an error because no configuration is provided.

Reviewer: Lizz, I disagree with `tilt up: [...] don't persist across sessions` because in my experience that's not how most dev apps behave nowadays. To take VSCode as an example:

I get VSCode up with `code -n ~/code/project`, and open a bunch of files in it. At the end of the day, I turn off the computer. Monday I get back to work, click the VSCode icon, and everything's how I left it. Maybe I decide to work on a different project so I run `code -r ~/code/different` which replaces whatever I had open there with the new project. Once I'm done with that, I then Ctrl+Shift+O and load up let's say `~/code/another` via VSCode's UI. I close VSCode, and when I open it again tomorrow morning, it shows me `~/code/another`, the last place I was at.

So I'd rather follow convention and have Tilt mimic that. It might not be possible since there's a lot of extra complexity, but I think at least `tilt up` with no arguments should.

A simpler solution which might just end the whole discussion could be:

```
$ tilt up
Continue where you left off [Y/n]?
```

`Y` shows the user the exact state of the last session, regardless of config file changes. `N` brings up a clean slate, regardless of what the previous state was. `tilt up` with extra arguments behaves as Nick described.

Whether this prompt appears or not, whether it's `[Y/n]` or `[y/N]`, and whether it's `Y` or `N` by default without prompting, can be configured via a `persist_previous_state()` setting on the Tiltfile.

Thoughts?

Reviewer 4: I love this idea - allow the user to decide. Han and I were working off a similar concept.. here's a sneak preview of that:

![Screen Shot 2021-10-01 at 1 58 31 PM](https://user-images.githubusercontent.com/46793797/135666099-e0a282ca-84f9-469b-8369-1b02f9f7cfe0.png)

Reviewer: Another idea, from Nick: Always restart from where the user left off, and show a warning that tells the user how to have a clean slate instead if they wish.
